THINGS/things.glmsingle/code/glmsingle/README.md

```diff
@@ -0,0 +1,294 @@
+```bash
+DATADIR="cneuromod-things/THINGS"
+python GLMsingle_cleanmask.py --things_dir="${DATADIR}"
+```
+
+**Input**:
+- All 4 subject's ``*bold.nii.gz`` files, for all sessions (~36) and runs (6 per session) \
+(e.g., ``sub-03_ses-10_task-things_run-1_space-T1w_desc-preproc_part-mag_bold.nii.gz``)
+- ``sub-{sub_num}_task-things_space-T1w_desc-func-union_mask.nii``, the
+functional mask generated from the union of the functional masks of every run in Step 2.
+
+**Output**:
+- ``sub-{sub_num}_task-things_space-T1w_desc-NaNvals_mask.nii``, a mask that
+includes any voxel from the functional union mask with at least one normalized NaN score.
+- ``sub-{sub_num}_task-things_space-T1w_desc-func-clean_mask.nii``, a functional
+mask excludes any voxel with normalized NaN scores from the functional union mask.
+
+NOTE: sub-06 session 8, run 6 was corrupted (brain voxels misaligned with other fmriprepped runs). All final analyses were redone without that run.
+
+------------
+
+## Step 6. Compute noise ceilings on trial-unique betas
+
+Derive voxelwise noise ceilings from beta scores estimated with GLMsingle.
+
+The noise ceiling estimation is adapted from the [Natural Scene Dataset's datapaper methodology](https://www.nature.com/articles/s41593-021-00962-x).
+
+**Preliminary step**:\
+To leave out "blank" trials (trials with no recorded subject
+response) from noise ceiling computations, trialwise performance needs to be
+extracted. Run the ``behav_data_annotate.py`` script, as described under
+**Trial-Wise Image Ratings and Annotations** in the ``cneuromod-things/THINGS/things.behaviour`` README. Output is saved as ``cneuromod-things/THINGS/things.behaviour/sub-{sub_num}/beh/sub-{sub_num}_task-things_desc-annotation-per-trial_beh.tsv``.
+
+To compute noise ceilings, launch the following script for each subject:
+```bash
+DATADIR="cneuromod-things/THINGS"
+python GLMsingle_noiseceilings.py --things_dir="${DATADIR}" --sub_num="01"
+```
+
+**Input**:
+- A subject's ``TYPED_FITHRF_GLMDENOISE_RR.mat``, a single .mat file outputed by GLMsingle (model D) in Step 4, which contains trial-unique betas per voxel
+- ``task-things_desc-runlist.h5``, a single file with nested lists of valid runs per session for each subject created in Step 3.
+- A subject's ``sub-{sub_num}_task-things_sparsedesign.h5`` file created in Step 1.
+- A subject's ``sub-{sub_num}_task-things_space-T1w_desc-func-union_mask.nii`` and
+``sub-{sub_num}_task-things_space-T1w_desc-func-clean_mask.nii`` masks created in Steps 2 and 5, respectively.
+- A subject's ``cneuromod-things/THINGS/things.behaviour/sub-{sub_num}/beh/sub-{sub_num}_task-things_desc-annotation-per-trial_beh.tsv``, a single .tsv file per subject with trialwise performance metrics and image annotations created with the ``cneuromod-things/THINGS/things.behaviour/code/behav_data_annotate.py`` script in the above preliminary step.
+
+**Output**:
+- ``sub-{sub_num}_task-things_space-T1w_res-func_modelD_noise-ceilings.mat``,
```

Comment:
`_res-func` was my attempt at implementing templateflow conventions when I added masks to the timeseries extraction code, to differentiate between masks at the same res as anatomical vs functional scans. 

Model D is one of the 4 models from the GLMsingle toolbox, there are 4 sets of output data. D is the better one. I should specify that in the doc...  

It's only one set of values per voxel (noise ceilings), so a .h5 file felt like overkill for just one array, don't you think? (the other .h5 files have arrays per trial and per image, they're big). 

Isn't label-GM for grey matter masks? How about `sub-{sub_num}_task-things_space-T1w_model-gmlsingleD_stat-noiseCeilings_statmap.mat`?

Reply:
I added the `label-GM` as my understanding was that you only compute these values from timeseries in a mask, maybe `GM` is not what it is, but it is better to specify which subset of the brain you are using. Is that the union mask with NaNs excluded?

Reply:
changed to `sub-{sub_num}_task-things_space-T1w_model-fitHrfGLMdenoiseRR_stats-noiseCeilings_statmap.nii.gz` saved as brain volume to avoid any issue with matching vectorized output to the right binary mask.